iceoryx_utils/include/iceoryx_utils/cxx/string.hpp

```diff
@@ -312,7 +401,130 @@ class string
     /// @return a std::string with data equivalent to those stored in the string
     operator std::string() const noexcept;
 
+    /// @brief since there are two valid options for what should happen when appending a string larger than this'
+    /// capacity (failing or truncating), the fixed string does not support operator+=; use append for truncating or
+    /// unsafe_append for failing in that case
+    template <typename T>
+    string& operator+=(const T&) noexcept;
+
+    /// @brief appends a fixed string or string literal to the end of this. If this' capacity is too small for appending
+    /// the whole string (literal) the remainder of the characters are truncated.
+    ///
+    /// @param [in] TruncateToCapacity_t is a compile time variable which is used to make the user aware of the possible
+    /// truncation
+    /// @param [in] t is the fixed string/string literal to append
+    ///
+    /// @return reference to self
+    ///
+    /// @code
+    ///     string<5> fuu("cde");
+    ///     fuu.append(TruncateToCapacity, "fgahc");
+    /// @endcode
+    template <typename T>
+    typename std::enable_if<internal::IsCharArray<T>::value || internal::IsCxxString<T>::value, string&>::type
+    append(TruncateToCapacity_t, const T& t) noexcept;
+
+    /// @brief appends a fixed string or string literal to the end of this. The appending fails if the sum of both sizes
+    /// is greater than this' capacity.
+    ///
+    /// @param [in] fixed string/string literal to append
+    ///
+    /// @return true if the appending succeeds, otherwise false
+    template <typename T>
+    typename std::enable_if<internal::IsCharArray<T>::value || internal::IsCxxString<T>::value, bool>::type
+    unsafe_append(const T& t) noexcept;
+
+    /// @brief creates a substring containing the characters from pos until count; if pos+count is greater than the size
+    /// of the original string the returned substring only contains the characters from pos until size();
+    /// iox::cxx::nullopt is returned if pos is greater than the size of the original string;
+    ///
+    /// @param [in] pos is the position of the first character used for the substring
+    /// @param [in] count is the requested length of the substring
+    ///
+    /// @return an optional containing the substring, iox::cc::nullopt if pos is greater than the size of the original
+    /// string
+    iox::cxx::optional<string<Capacity>> substr(uint64_t pos, uint64_t count) const noexcept;
```

Comment:
I like the usage of `optional` here :)

Reply:
@MatthiasKillat the downside would be the following:
```
  std::string fuu = "hello world";
  std::string bla = fuu.substr(100000, 10);
  
  bla[4] = 'a'; // this is valid since substr did return a string of size 10 otherwise the stl would have thrown an error.
```
In our case `bla[4] = 'a';` it would lead to a segfault or even worse, to undefined behavior. Everytime I extract a string with substr I would have to test if the string contains the right number of characters which is more error prone then to have an optional which contains only then a string if one could be extracted.

Reply:
yes, please don't create surrogates if something can fail. It has to be documented and the user has to read the documentation. With the optional, it's clear that there might be no value inside